File: geaflow/geaflow-plugins/geaflow-store/geaflow-store-rocksdb/src/main/java/com/antgroup/geaflow/store/rocksdb/iterator/VertexScanIterator.java
/*
 * Copyright 2023 AntGroup CO., Ltd.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 */

package com.antgroup.geaflow.store.rocksdb.iterator;

import com.antgroup.geaflow.common.tuple.Tuple;
import com.antgroup.geaflow.model.graph.vertex.IVertex;
import com.antgroup.geaflow.state.iterator.IVertexIterator;
import com.antgroup.geaflow.state.pushdown.IStatePushDown;
import com.antgroup.geaflow.state.pushdown.filter.inner.IGraphFilter;
import java.util.Iterator;
import java.util.function.BiFunction;

public class VertexScanIterator<K, VV, EV> implements IVertexIterator<K, VV> {

    private final Iterator<Tuple<byte[], byte[]>> iterator;
    private final IGraphFilter filter;
    private final BiFunction<byte[], byte[], IVertex<K,VV>> vertexDecoder;
    private boolean isClosed = false;
    private IVertex<K, VV> nextValue;

    public VertexScanIterator(Iterator<Tuple<byte[], byte[]>> iterator,
                              IStatePushDown pushdown,
                              BiFunction<byte[], byte[], IVertex<K, VV>> decoderFun) {
        this.vertexDecoder = decoderFun;
        this.iterator = iterator;
        this.filter = (IGraphFilter) pushdown.getFilter();
    }

    @Override
    public boolean hasNext() {
        if (isClosed) {
            return false;
        }
        while (iterator.hasNext()) {
            Tuple<byte[], byte[]> pair = iterator.next();
            nextValue = vertexDecoder.apply(pair.f0, pair.f1);

            if (!filter.filterVertex(nextValue)) {
                continue;
            }
            return true;
        }

        return false;
    }

    @Override
    public IVertex<K, VV> next() {
        return nextValue;
    }
}

File: geaflow-cstore/src/api/iterator/vertex_scan_iterator.rs
// Copyright 2023 AntGroup CO., Ltd.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.

use std::{iter::Peekable, rc::Rc};

use crate::{
    api::{
        filter::{
            store_and_filter::StoreAndFilter, store_only_vertex_filter::StoreOnlyVertexFilter,
            EdgeLimit, FilterHandler, StoreFilterPushdown,
        },
        graph::{graph_serde::GraphSerde, vertex::Vertex},
    },
    iterator::scan_data_iterator_with_edge_limit::ScanDataEdgeLimitedIterator,
    Engine,
};

/// Return VertexScanIter by calling scan interface.
pub struct VertexScanIter<'a> {
    graph_serde: &'a GraphSerde,
    engine: &'a Engine,
    scan_data_merge_iterator: Peekable<ScanDataEdgeLimitedIterator<'a>>,
}

impl<'a> VertexScanIter<'a> {
    pub fn new(
        graph_serde: &'a GraphSerde,
        engine: &'a Engine,
        filter_pushdown: StoreFilterPushdown,
    ) -> Self {
        Self {
            graph_serde,
            engine,
            scan_data_merge_iterator: engine.scan(Rc::new(VertexScanIter::generate_new_pushdown(
                filter_pushdown,
            ))),
        }
    }

    fn generate_new_pushdown(filter_pushdown: StoreFilterPushdown) -> StoreFilterPushdown {
        let only_vertex_filter = FilterHandler::OnlyVertex(StoreOnlyVertexFilter::default());

        let mut filter_vec = vec![filter_pushdown.filter_handler];
        filter_vec.push(only_vertex_filter);

        let new_filter_handler = FilterHandler::And(StoreAndFilter::new(filter_vec));
        StoreFilterPushdown::new(
            new_filter_handler,
            filter_pushdown.filter_context,
            EdgeLimit::default(),
        )
    }
}

impl<'a> Iterator for VertexScanIter<'a> {
    type Item = Vertex;

    fn next(&mut self) -> Option<Self::Item> {
        if let Some(merged_graph_data) = self.scan_data_merge_iterator.next() {
            return Some(self.graph_serde.deserialize_vertex_in_scan(
                &merged_graph_data.graph_data,
                &self.engine.label_dict,
            ));
        }

        None
    }
}
